### ISSUE TITLE
Integration reports sensor as unavailable sporadically and floods log

### ISSUE THREAD

**Issue by KennethLavrsen** [agittins/bermuda#85]
## Version of the custom_component
Version 0.4.3

## Configuration

```yaml
I have one device configured via UI and you do not use YAML for config
```

## Describe the bug

I have just setup two ESPHome BT Proxies and I have no local BT device on HA. I have one noname small BLE iBeacon device.

At first the integration seems to work great and as expected. But then the device suddenly appear unknown and when that happens the HA log gets flooded as seen below

## Debug log

```text

2024-01-09 21:34:46.850 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: f6:0a:34:1a:0d:03 10dB
2024-01-09 21:34:46.852 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: cb:d1:5c:2e:17:0e -127dB
2024-01-09 21:34:51.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: f6:0a:34:1a:0d:03 -127dB
2024-01-09 21:34:51.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: fe:1f:3b:98:cc:e6 -127dB
2024-01-09 21:34:56.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: f6:0a:34:1a:0d:03 10dB
2024-01-09 21:34:56.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: fe:1f:3b:98:cc:e6 10dB
2024-01-09 21:34:56.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: dd:7a:8c:77:be:df 10dB
2024-01-09 21:34:56.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: cb:d1:5c:2e:17:0e 10dB
2024-01-09 21:35:01.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: f6:0a:34:1a:0d:03 -127dB
2024-01-09 21:35:01.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: dd:7a:8c:77:be:df -127dB
2024-01-09 21:35:01.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: cb:d1:5c:2e:17:0e -127dB
2024-01-09 21:35:06.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: d2:18:8c:f7:4f:93 -127dB
2024-01-09 21:35:06.653 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: dd:7a:8c:77:be:df 10dB
2024-01-09 21:35:11.651 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: d2:18:8c:f7:4f:93 10dB
2024-01-09 21:35:11.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: dd:7a:8c:77:be:df -127dB
2024-01-09 21:35:16.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: d2:18:8c:f7:4f:93 -127dB
2024-01-09 21:35:16.653 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: dd:7a:8c:77:be:df 10dB
2024-01-09 21:35:16.653 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: e7:b6:6a:84:73:bc -127dB
2024-01-09 21:35:16.653 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: cb:d1:5c:2e:17:0e 10dB
2024-01-09 21:35:21.651 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: d2:18:8c:f7:4f:93 10dB
2024-01-09 21:35:21.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: f6:0a:34:1a:0d:03 10dB
2024-01-09 21:35:21.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: e7:b6:6a:84:73:bc 10dB
2024-01-09 21:35:26.654 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: dd:7a:8c:77:be:df -127dB
2024-01-09 21:35:26.655 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: e7:b6:6a:84:73:bc -127dB
2024-01-09 21:35:26.655 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: cb:d1:5c:2e:17:0e -127dB
2024-01-09 21:35:31.653 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: dd:7a:8c:77:be:df 10dB
2024-01-09 21:35:31.653 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: e7:b6:6a:84:73:bc 10dB
2024-01-09 21:35:36.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: f6:0a:34:1a:0d:03 -127dB
2024-01-09 21:35:41.653 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: d2:18:8c:f7:4f:93 -127dB
2024-01-09 21:35:41.654 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: fe:1f:3b:98:cc:e6 -127dB
2024-01-09 21:35:41.655 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: dd:7a:8c:77:be:df -127dB
2024-01-09 21:35:46.653 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: d2:18:8c:f7:4f:93 10dB
2024-01-09 21:35:46.654 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: f6:0a:34:1a:0d:03 10dB
2024-01-09 21:35:46.655 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: fe:1f:3b:98:cc:e6 10dB
2024-01-09 21:35:51.652 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: f6:0a:34:1a:0d:03 -127dB
2024-01-09 21:35:51.653 WARNING (MainThread) [custom_components.bermuda] Device changed TX-POWER! That was unexpected: cb:d1:5c:2e:17:0e 10dB

```

**Comment by KennethLavrsen** [comment]
I have put back the integration.

I will add updates step by step.

It is the both distance and area sensors that occationally go to unknown. The device_tracker stays at "home" as it shoud. The unknown lasts only 1-2 seconds. It happens once a minute. There are no error logged for this.

I have monitored the log and the sensors do not always go unknown even though the log gets populated rapidly (I call that flood but English is not my first language so maybe it is too harch). It adds 30 lines to the log per minute. That is a lot when you try to find if there are other problems with the installation. I will downgrade it to debug by hacking the code. I can see what you did on github.

These iBeacons are powered by small 2032 coin cells.  Maybe the integration needs to expect a slower update rate from iBeacons than you have from devices with much more battery capacity like phones.
These beacons do not seem to change MAC. They are made to be recognized and tracked unlike iPhones.
It is my subjective and with no proof "feeling" that the integration is not patient enough before it declares it unknown. And we are talking seconds.

These tags are purchased on Aliexpress. They look like this. I printed the plastic housing myself

![IMG_4797](https://github.com/agittins/bermuda/assets/8373419/f673c542-3261-4b04-b378-a77372d21137)

I will return with more that you asked for later

**Comment by KennethLavrsen** [comment]
I have some more data. I put a beacon in the kitchen and one temporarily in my office where 4 computers were on. It is interference that jams the one in the office. The kitchen one does not change to unknown and its distance jumps between 0.4 and 1 meter (it is 0.3 m). The ofiice one (named living room) jumps between several meters and 0.5. Interference cannot be avoided. Bluetooth uses frequency hopping so interference is not constant. So you may consider a time out of a few seconds more before putting the state to unknown. That will help people doing automations based on area to avoid having to manage frequent unknown states

**Comment by KennethLavrsen** [comment]
I realize that when I raised this bug I made a huge mistake. I wrote Unknown. The state is unavailable.

So to wrap up. There are two issues and they are totally independent.

1. the logging relates to bluetooth devices in my house that I never intended to track. Many are probably Hue light bulbs that run Zigbee and just polite the radio spectrum with bluetooth. Lowering the debug level to debug does it for me.

2. the unavailable issue. This is surely interference. After I relocated the proxy and put the iBeacon closer the states are stable and known now for 8 hours. If the interference is massive then unavailable is the right state. Otherwise we cannot recognize when the beacon leaves the house. But if the interference is just packets that are lost when the frequency hopping hits a wifi or thread or zigbee transmission, which is seen as unavailable for 1-2 seconds in the UI then I think the user is better off that the integration has a little hangtime before it reports unavailable. Or increase it if it already has one. An iBeacon should transmit at least one per second so we are taking very few seconds hangtime. Is there already one? If so I can experiment with increasing it

**Comment by agittins** [comment]
The logging won't be caused by zigbee packets, it's literally a bluetooth device sending an advertisement that includes a value specifying a certain power level, then sending another one where it says it is using a different tx power level - it's not interference in the logging case but the deliberate behaviour of some bluetooth devices. But that's solved for you now so we'll concentrate on the `unavailable` part.

tldr: Your `Maximum Radius` setting in Bermuda might be too low. Try 30 metres or more.

There's a few things to consider.
- As you mention, bluetooth is using the shared 2.4GHz ISM spectrum, so we _expect_ it to be an unreliable mechanism. Lost packets are normal and expected, and radio propagation above 900MHz is tightly bound with occult practices.
- A single BLE Advertisement at the RF layer is typically a given packet, sent on the three advertising channels one after the other (so in the order of <6ms or so - there might be an exact number for that).
- Devices may send an advertisement at any interval from 20ms to 10.24 _seconds_. (Interestingly, devices with advertising intervals higher than 1.8seconds are really difficult to connect to in linux)
- A receiver like the ble proxy needs to decide how to listen, since it can only listen on one bluetooth channel at a time, and also in the case of ESP32 devices that are using Wifi, it also has to spend some of that time listening/sending wifi traffic instead, as they only have a single radio in them.

Soo... this leads us to:

- What configuration are you using on the two ESPHome proxies? Specifically are you using defaults for `interval` and `window` under:
```yaml
esp32_ble_tracker:
  scan_parameters:
    interval: 320ms
    window: 30ms
```
The above values are the defaults if you don't specify anything[^esphomeble]. The interval is the time _between_ each window, and a window is the amount of time it spends listening to _a single ble channel_.

So we have three advertising channels (37, 38, 39) [^bleadv], and with the default values the ESP32 will spend 30ms listening on channel 37, then _not listen_ for 290ms (it can do other tasks and wifi comms in that period), then listen to channel 38 for 30ms and so-on. So in bluetooth terms you get something like:

```
Time--->>>
|RX(37)|-----------------Deafness---------------|RX(38)|---------------Deafness---------------|RX(39)|----
```

Obviously this means it's going to miss A LOT of packets - since it's not even listening for 90% of the time! Transmitters will add a random delay between 0 and 10ms to each advertising interval, to ensure that devices don't (always) collide, and likewise the advertising interval should slide over the listening window over time.

This is ok-ish for things like a thermometer that advertises every second but the data stays relevant for many seconds, or a device that one intends to connect to and waiting a second or two for it to show up in a list is OK. But when we want to track things moving around (or if you have a BLE device sending a Switch command or  other transient event) it becomes more important to lose fewer packets.

In my experimentation I have found that configuring the ESP with an `interval` of 1000ms and a `window` of 900ms is "pretty good", I get packets at an interval of 2 to 4 seconds typically, perhaps up to 6 seconds if I'm unlucky (this is for my watch which transmits at ~ 1s intervals). If you set the `window` too close to the `interval` then the ESP won't have enough time to perform wifi tasks, and will start losing its connection, but any gap between them is dead-time that you're not listening for bluetooth packets. Using a large interval  increases one's chances of hearing every advert at least once in a single window, but the larger the interval the more likely some wifi stuff is to time out, or the amount of data (due to received BLE adverts) that needs to be sent via wifi might be too much to send in the available dead-time.

Of course an esp32 with ethernet could run with much tighter timings for minimum dead-time and quicker cycling between channels and quicker reports upstream (but the defaults aren't set up for that either). Bottom line: experimenting with `window` and `interval` is very worthwhile, but beware that as soon as you get away from the very conservative defaults your odds of not being able to connect to it via wifi (say, in order to flash it back to sensible values) might be unlikely. Flashing via usb/serial will always get you back, though.

I know that might be a lot more background than desired, but it's pretty important in order to understand what we're up against and the various places that our data likes to disappear on us!

So for a bit more info, I'd suggest:
- Install NRFConnect on a phone, and use it to show you what the advertising interval is from your beacon. Eg, here's what it shows for my watch and a generic BLE Beacon tag:  
![image](https://github.com/agittins/bermuda/assets/752405/cd70212b-75a0-46b2-98a0-673c5499cc37)

- Use the `dump_devices` service in Bermuda to look at what the interval history looks like for that device. Eg: 
![image](https://github.com/agittins/bermuda/assets/752405/93a3df7c-2667-4e02-8837-e1c146486908)

- What is your Bermuda configuration? If you have `CONF_MAX_RADIUS` set to something small then it would be correct for the beacon to be marked `Unknown` if the estimated range is beyond that distance. If you want it to always read as near it's closest proxy regardless of how near, set it large. I use 30m: 
![image](https://github.com/agittins/bermuda/assets/752405/f62027ad-860f-44d5-9fb8-bc7eb760941d)


I suspect that your max radius might be too small, because the way it should work is that we persist the closest measurement regardless of age, we only consider the freshness of the advert when choosing between two proxies, in which case a newer advert wins even if an older one is closer (within a certain age) - this will likely be modified soon, but the max_radius setting is what's probably hitting you right now.

[^bleadv]: https://academy.nordicsemi.com/courses/bluetooth-low-energy-fundamentals/lessons/lesson-2-bluetooth-le-advertising/topic/advertising-process/
[^esphomeble]: https://esphome.io/components/esp32_ble_tracker.html#configuration-variables

**Comment by agittins** [comment]
By the way I like the printed key-fob case for the module! Looks sturdy but space-efficient.

**Comment by KennethLavrsen** [comment]
Thanks for a very detailed follow up. The reference to Hue bulbs and the logging is that the hue bulbs are zigbee but they also run bluetooth and annouce their presence for those that want to control them by BT. I cannot imagine why anyone would do that. I would love to be able to turn the BT radio off in these bulbs but you cannot.

I will play with the settings tonight when I am home from work.

The BT proxy that give the Unavailable is actually the Ethernet wired one. It is using this setting

```
esp32_ble_tracker:
  scan_parameters:
    interval: 1100ms
    window: 1100ms
    active: true
```
which is default for the wired Olimax ESP installed via https://esphome.io/projects/?type=bluetooth

I will try and experiement as you suggests

**Comment by KennethLavrsen** [comment]
Hi. I think I have some important info now - following your advice

I never touched any of the Bermuda defaults. I only added my two beacons. So the Max radius for simple Area detection was 3 meters. I increased that to 6 meters. And I moved the beacon to a distance of 1.1 meters instead of the 20 cms.

I have seen several reports of the distance peaking to above 3 meters. So that explains the Unavailable (which is actually shown as Unknown in the graphical presentation in HA so I was not complete idiot when I reported it).  So mayabe the integration should install with some larger defaults?

We agree that the logging is resolved with the downgrade to debug. But I know you really want to know what devices are reporting the TX power jumps. I installed the NRF Connect on an Android (I had it on iPhone but it does not allow to show MAC it seems). And the result is..... Tada.

ALL are Hue Bulbs. All. I did not find a single example that was not. The Hue bulbs create this spectrum polution. Yes the bluetooth is a feature so a newbie can pretend he has a smarthome by turning a few bulbs on and off using the Hue app with just BT. And yes it helps provisioning the bulbs. But I wish they would turn the Bluetooth radio off once they are paired with a Zigbee coordinator. But we cannot fix Philips Hue. We have to live with it.  But that is the root of those strange transmissions so now you know.

In summary

1. Logging fixed by downgrading to debug
2. The unavailable could be improved by shipping a larger default value for the area detect radius. But if the beacon really is too far away - the state is correct. It is unknown then.

**Comment by agittins** [comment]
Thanks for following up on this.

Good to know that it was the hue bulbs in your case that have the varying tx power. 

The default for max radius has been 20 metres since about v0.3.1 in October, (it was 3m prior to that) so perhaps you had previously installed an older version and had an older default from back then? Since the current default is 20m that should take care of "most" installations. For those wanting a more reliable "assertion" that a device has left the area reducing it could be helpful, but I won't worry about tuning the default further until I do more work on the selection algo.

Glad we got to the bottom of it! I'll close this out now since new installs should be getting the 20m max radius by default.

**Comment by KennethLavrsen** [comment]
Thanks. Maybe I changed the value to 3 m and forgot. I am only human. Thanks for all the help and for this cool integration. Adding my star
